src/light_data/light_store.rs

```diff
@@ -0,0 +1,405 @@
+                LightData::Cell(vec![LightData::Atom(vec![3u8]), FWrap(*f).ser(), x.ser()])
+            }
+            LightExpr::Num(f) => LightData::Cell(vec![LightData::Atom(vec![]), FWrap(*f).ser()]),
+            LightExpr::Char(f) => LightData::Cell(vec![LightData::Cell(vec![]), FWrap(*f).ser()]),
+            LightExpr::Nil => LightData::Atom(vec![0u8]),
+            LightExpr::StrNil => LightData::Atom(vec![1u8]),
+            LightExpr::SymNil => LightData::Atom(vec![2u8]),
+        }
+    }
+    fn de(ld: &LightData) -> anyhow::Result<Self> {
+        match ld {
+            LightData::Atom(v) => match v[..] {
+                [0u8] => Ok(LightExpr::Nil),
+                [1u8] => Ok(LightExpr::StrNil),
+                [2u8] => Ok(LightExpr::SymNil),
+                _ => Err(anyhow!("LightExpr::Atom({:?})", v)),
+            },
+            LightData::Cell(v) => match &v[..] {
+                [LightData::Atom(u), ref x, ref y] => match u[..] {
+                    [0u8] => Ok(LightExpr::Cons(ScalarPtr::de(x)?, ScalarPtr::de(y)?)),
+                    [1u8] => Ok(LightExpr::StrCons(ScalarPtr::de(x)?, ScalarPtr::de(y)?)),
+                    [2u8] => Ok(LightExpr::SymCons(ScalarPtr::de(x)?, ScalarPtr::de(y)?)),
+                    [3u8] => Ok(LightExpr::Comm(FWrap::de(x)?.0, ScalarPtr::de(y)?)),
+                    _ => Err(anyhow!("LightExpr::Cell({:?})", v)),
+                },
+                [LightData::Cell(u), ref x] => match u[..] {
+                    [] => Ok(LightExpr::Char(FWrap::de(x)?.0)),
+                    _ => Err(anyhow!("LightExpr::Cell({:?})", v)),
+                },
+                [LightData::Atom(u), ref x] => match u[..] {
+                    [] => Ok(LightExpr::Num(FWrap::de(x)?.0)),
+                    _ => Err(anyhow!("LightExpr::Cell({:?})", v)),
+                },
+                _ => Err(anyhow!("LightExpr::Cell({:?})", v)),
+            },
+        }
+    }
+}
+
+#[cfg(test)]
+pub mod tests {
+    use crate::parser;
+
+    use super::*;
+    use pasta_curves::pallas::Scalar;
+    use std::collections::BTreeMap;
+
+    proptest! {
+        #[test]
+        fn prop_light_expr(x in any::<LightExpr<Scalar>>()) {
+            let ser = x.ser();
+            let de  = LightExpr::de(&ser).expect("read LightExpr");
+            assert_eq!(x, de)
+        }
+
+        #[test]
+        fn prop_light_store(s in any::<LightStore<Scalar>>()) {
+            let ser = s.ser();
+            let de  = LightStore::de(&ser).expect("read LightStore");
+            assert_eq!(s, de)
+        }
+
+        #[test]
+        fn test_convert_light_store_basic_strings((ptr3, ptr4, c1, c2) in any::<(ScalarPtr<Scalar>, ScalarPtr<Scalar>, char, char)>().prop_filter(
+            "Avoids confusion with StrNil",
+            |(ptr3, ptr4,c1, c2)| {
+                let strnil = ScalarPtr::from_parts(ExprTag::Str.as_field(), Scalar::zero());
+                *ptr3 != strnil && *ptr4 != strnil && *c2 != '\0' && *c1 != '\0'
```

Comment:
What's the reason for `*c2 != '\0' && *c1 != '\0'`? Afaiu ascii null is a valid/representable char, and can be in Lurk strings, no?

Reply:
No, you're correct, that works. I had my doubts about the behavior of the `.chars()` enumerator used in `insert_scalar_string`: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=c5fe4c9e522c867da1e03bf4ba59e74e
I'll fix.